### PR TITLE
Add overlay for mesh config to helm

### PR DESCRIPTION
**Pull request by howardjohn**
A similar API exists in the operator. This does this in native Helm,
allowing it to be used directly with Helm and to avoid re-implementing
the functionality in operator.

There are two motivations for this change:

* Users only need to learn one API, which has been around forever and is
revewed in istio/api: MeshConfig. They have direct access to modify
this, rather than an indirection, where we add confusion and less
flexibility. With this change, users can modify any setting they desire.
* Zero maintainance. We don't need to add new values for every field, as
users can touch all of them. We can also remove all the values here (in
the future), as they are redundant now - this cleans up ~60 values.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -19,12 +19,12 @@ require (
 	fortio.org/fortio v1.3.1
 	github.com/Azure/go-autorest/autorest v0.9.4 // indirect
 	github.com/DataDog/datadog-go v2.2.0+incompatible
+	github.com/Masterminds/goutils v1.1.0 // indirect
 	github.com/Masterminds/semver v1.4.2
-	github.com/Masterminds/sprig v2.14.1+incompatible
+	github.com/Masterminds/sprig v2.20.0+incompatible
```
Comment (howardjohn) [inline go.mod]: We needed a sprig update to pick up `mergeOverwrite`

**Comment by howardjohn** [comment]
I have a more complete version of this in https://github.com/istio/istio/pull/22593, but its broken for some reason - not sure what.

File: manifests/istio-control/istio-discovery/templates/configmap.yaml [hunk 2]
```diff
@@ -287,6 +241,40 @@ data:
 {{ toYaml .Values.global.proxy.envoyAccessLogService.tcpKeepalive | trim | indent 10 }}
       {{- end}}
     {{- end}}
+{{- end }}
+
+{{/* We take the mesh config above, defined with individual values.yaml, and merge with .Values.meshConfig */}}
```
Comment (costinm) [inline manifests/istio-control/istio-discovery/templates/configmap.yaml]: Why do we need the values above ? Do we need all of them ?

Comment (costinm) [reply manifests/istio-control/istio-discovery/templates/configmap.yaml]: I mean - are they 'default values' ? How many match the default values in code ( and can we fix the code to have the proper default values and do the 'if' as needed ) ?

Comment (howardjohn) [reply manifests/istio-control/istio-discovery/templates/configmap.yaml]: For backwards compatibility. I think we should leave them around for 1.6, and have it print a warning when they are used. Then in 1.7 (or really, in a few weeks when we cut the branch) we remove. What do you think?

File: manifests/istio-control/istio-discovery/values.yaml [hunk 3]
```diff
@@ -200,3 +200,7 @@ telemetry:
 
 # Revision is set as 'version' label and part of the resource names when installing multiple control planes.
 revision: ""
+
+# meshConfig defines runtime configuration of components, including Istiod and istio-agent behavior
+# See https://istio.io/docs/reference/config/istio.mesh.v1alpha1/ for all available options
+meshConfig: {}
```
Comment (costinm) [inline manifests/istio-control/istio-discovery/values.yaml]: Sounds great - can you confirm we're not using and global. in the template ? 
Global is a big problem - last artifact of the spaghetti templates...

File: manifests/istio-control/istio-discovery/templates/configmap.yaml (hunk 2 above)
Comment (howardjohn) [reply manifests/istio-control/istio-discovery/templates/configmap.yaml]: The defaults vs not defaults isn't really the issue here, its more about the "api" I think

File: manifests/istio-control/istio-discovery/values.yaml (hunk 3 above)
Comment (howardjohn) [reply manifests/istio-control/istio-discovery/values.yaml]: Yes, no `.global` here, its configured like `--set meshConfig.discoveryAddress=foo`. Is that what you mean?

**Review by costinm — APPROVED** [review]
Clearly an improvement. 

But the biggest decision really is if we even want installer to create the configmap - instead of using it 
as a runtime config ( including the long requested conversion to a CRD ). 

If the code-based defaults are made to match the default install - I think we can do this. 

I would also much prefer the 'merge' to happen in code, not helm template: i.e. create a second
config map with whatever is in values.yaml, and have the code read that and merge with possibly 
another user-edited configmap and the default values.

But this can be done later - I think this PR is on the right path.

**Pull request merged**